File: 445.两数相加-ii.cpp
/*
 * @lc app=leetcode.cn id=445 lang=cpp
 *
 * [445] 两数相加 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {

        // push
        while (l1)
        {
            stack1.push(l1->val);
            l1 = l1->next;
        }
        while (l2)
        {
            stack2.push(l2->val);
            l2 = l2->next;
        }
        int c = 0;
        ListNode *ans = NULL;
        while (!stack1.empty() || !stack2.empty() || c != 0)
        {
            int a = stack1.empty() ? 0 : stack1.top();
            int b = stack2.empty() ? 0 : stack2.top();
            if (!stack1.empty())
                stack1.pop();
            if (!stack2.empty())
                stack2.pop();
            int current = a + b + c;
            c = current / 10;
            current %= 10;
            ListNode *curcode = new ListNode(current);
            curcode->next = ans;
            ans = curcode;
        }
        return ans;
    }

private:
    stack<int> stack1, stack2;
};
// @lc code=end

File: JavaScript/445.两数相加-ii.js
/*
 * @lc app=leetcode.cn id=445 lang=javascript
 *
 * [445] 两数相加 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  l1 = reverseList(l1);
  l2 = reverseList(l2);

  let carry = 0;
  let head = null;

  while (l1 || l2 || carry) {
    let sum = carry;

    if (l1) {
      sum += l1.val;
      l1 = l1.next;
    }

    if (l2) {
      sum += l2.val;
      l2 = l2.next;
    }

    carry = Math.floor(sum / 10);
    let digit = sum % 10;

    let newNode = new ListNode(digit);
    newNode.next = head;
    head = newNode;
  }

  return head;
};
function reverseList(head) {
  let prev = null;
  let curr = head;

  while (curr) {
    let next = curr.next;
    curr.next = prev;
    prev = curr;
    curr = next;
  }

  return prev;
}
// @lc code=end
